### PR TITLE
ENH: add hash check

### PR DESCRIPTION
**Pull request by nrwslac**
Hash the atef config file and clean up some other bits of code. I tried to add the signal docs in the cpt for later access, but found that difficult.

### CODE REVIEW

File: __main__.py [hunk 1]
```diff
@@ -11,27 +11,45 @@
...
     print('Missing options: %s' % ' '.join(missing))
     sys.exit(-1)
 
+
 def SetupOptionParser():
     parser = optparse.OptionParser(usage=__doc__)
     parser.add_option('--run_atefics',
                     action='store_true',
                     dest="run_atefics",
                     help='run an atef check on a configuratin file')
```
Comment (tangkong) [inline __main__.py]:
```suggestion
                    help='run an atef check on a configuration file')
```

File: __main__.py [hunk 2]
```diff
@@ -11,27 +11,45 @@
 from email.utils import make_msgid
 import optparse
 
-client_id_path = "./atefics/client.json"
+client_id_path = "/reg/g/pcds/epics-dev/nrw/ATEF/atefics/client.json"
```
Comment (tangkong) [inline __main__.py:15]: This is the json with the refresh token correct?  Even better might be allowing the user to add a path as a cli argument.  But doing that means you also probably want to parametrize the sender / receiver email addresses.  For now this is probably fine but food for thought.

File: __main__.py [hunk 3]
```diff
@@ -11,27 +11,45 @@
...
+devices = ['SP1K1 MONO', 'MR1K1 BEND', 'MR2K2 FLAT', 'MR3K2 KBH']
+
+def FilePathUtil(config_file):
+    if os.path.isfile(config_file):
+        return config_file
+    elif os.path.isfile("./atefics/checkouts/" + config_file):
+        return "./atefics/checkouts/" + config_file
+    else:
+        print(f"{config_file} does not exist")
+        sys.exit(-1)
```
Comment (tangkong) [inline __main__.py]: The more "pythonic" thing to do is raise an exception (maybe `FileNotFoundError`), then either let that stop your program or catch it higher up and return.  But nothing strictly wrong with this

File: __main__.py [hunk 4]
```diff
@@ -58,15 +77,32 @@ def refresh_token():
...
+    msg['to'] = ", ".join(recipients)
+    msg['Content-Type'] = "text/plain"
+    content = ""
+
+    with open(client_id_path, 'r') as fd:
+        client_details = json.loads(fd.read())
+    current_hash = hash_config(config_file)
+    saved_hash = client_details['web']['rix_config_hash']
+
+    if current_hash != saved_hash:
```
Comment (tangkong) [inline __main__.py:96]: A bit of a more existential question: why are we doing this?  Is this config never going to update?  Are you afraid of someone accidentally modifying your config file?

On the technical side, if you're looking to compare config files I think you can compare them in python with `==`, but having the hash in the email is nice too.

**Review by tangkong — COMMENTED** [review]
some quick thoughts

**Comment by tangkong** [comment]
> I tried to add the signal docs in the cpt for later access, but found that difficult.

Could you elaborate on this?  are you trying to get the `doc` field from the device's signal into the `Comparison` description field?  If so you'd access it from the device class itself, not the instantiated device

```python
OpticsHard.pos_lag_time.doc
```

File: __main__.py (hunk 2 above)
Comment (tangkong) [reply __main__.py:15]: You might also put the sender/receiver email in the client json itself and just make that one big bundle of information.  Just thinking about how we might generalize this for other people to use.

**Comment by nrwslac** [comment]
> > I tried to add the signal docs in the cpt for later access, but found that difficult.
> 
> Could you elaborate on this? are you trying to get the `doc` field from the device's signal into the `Comparison` description field? If so you'd access it from the device class itself, not the instantiated device
> 
> ```python
> OpticsHard.pos_lag_time.doc
> ```

no-doc
I would have thought the same, but here it says that doc isnt an attribute?

**Comment by nrwslac** [comment]
Whats more is, I thought there should be a way to access the .DESC field of the PV through the signal, but couldn't find a path for that either.

**Comment by tangkong** [comment]
Ah so you need to access the class, not the instance.  As in:
```python
hardstop_config = OpticsHard(base_pv, name='')

print(OpticsHard.pos_lag_time.doc)  # works
print(hardstop_config.pos_lag_time.doc)  # fails
```

the .DESC field is something you'd need to have a specific component for.  Many devices have it, but if you're just using EpicsSignal it won't know the .DESC field is there (in other words it won't assume the .DESC field exists)

**Comment by nrwslac** [comment]
> Ah so you need to access the class, not the instance. As in:
> 
> ```python
> hardstop_config = OpticsHard(base_pv, name='')
> 
> print(OpticsHard.pos_lag_time.doc)  # works
> print(hardstop_config.pos_lag_time.doc)  # fails
> ```
> 
> the .DESC field is something you'd need to have a specific component for. Many devices have it, but if you're just using EpicsSignal it won't know the .DESC field is there (in other words it won't assume the .DESC field exists)

I see. If I had two instances of OpticsHard, how would it know which one I was referencing?

**Comment by tangkong** [comment]
It wouldn't, but the signal component docs are defined in the class, so each instance would be the same.

**Comment by nrwslac** [comment]
> It wouldn't, but the signal component docs are defined in the class, so each instance would be the same.

Is this an OOP anti-pattern?

**Comment by tangkong** [comment]
Not necessarily.  This is equivalent to the docstring of every class instance being the same. Unless there's a specific antipattern you're looking at that I don't know about

**Comment by nrwslac** [comment]
> > Ah so you need to access the class, not the instance. As in:
> > ```python
> > hardstop_config = OpticsHard(base_pv, name='')
> > 
> > print(OpticsHard.pos_lag_time.doc)  # works
> > print(hardstop_config.pos_lag_time.doc)  # fails
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > the .DESC field is something you'd need to have a specific component for. Many devices have it, but if you're just using EpicsSignal it won't know the .DESC field is there (in other words it won't assume the .DESC field exists)
> 
> I see. If I had two instances of OpticsHard, how would it know which one I was referencing?

Nvm, I think I just didn't see in the code where this was defined as a class variable and not an instance variable. I'll need to look again.

File: __main__.py (hunk 2 above)
Comment (nrwslac) [reply __main__.py:15]: Maybe the more permanent recipients get added to the client.json and then we add an option to pass a list of additional recipients, and maybe an option to bypass the recipients in the client.json? I am not sure how many more features we should add do this until its no longer just a prototype.

Comment (tangkong) [reply __main__.py:15]: Yeah my suggestions are take it or leave it.  This one in particular is more of a future enhancement idea, but one that we'd need to make this more widely usable.

File: __main__.py (hunk 1 above)
Comment (nrwslac) [reply __main__.py]: fixed.

File: __main__.py (hunk 3 above)
Comment (nrwslac) [reply __main__.py]: updated.

File: __main__.py (hunk 4 above)
Comment (nrwslac) [reply __main__.py:96]: This is to ensure that the config doesnt change. That is, we want to guarantee we are checking the current values vs some unchanged values for radiation protection purposes. 
We could compare config files, but then we'd need to save an "original copy" of the file to compare it to.

**Comment by nrwslac** [comment]
In the end I found a pretty convenient way to iterate through all the signals and get the values needed using the getatt() fcn. This reduced the big clunky if else statements to a few lines. I should work on parsing the base_pv next.

File: __main__.py (hunk 4 above)
Comment (tangkong) [reply __main__.py:96]: I guess I thought that having an "original copy" that's read only would be helpful later down the line, but this is fine too.

**Comment by tangkong** [comment]
I think this is good to merge, I think for prototyping my reviews should be optional.  If you want a green stamp though I can do that

**Comment by nrwslac** [comment]
> I think this is good to merge, I think for prototyping my reviews should be optional. If you want a green stamp though I can do that

I appreciate the feedback. I want to double check that the new way I generate the config file worked as I expected and then I'll be ready to merge.

**Pull request merged**